### PR TITLE
Tdl 16328 implement request timeout

### PR DESCRIPTION
**Pull request by namrata270998**
Description of change
Added request timeout for SDK library functions with default timeout 300 seconds

Manual QA steps

- The request should backoff 5 times in case it is timed out.
- Error should be thrown in case of retries not successful after 5 trials
- Checked the request timeout works for each stream and backoff

# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_dynamodb/dynamodb.py [hunk 1]
```diff
@@ -66,16 +68,40 @@ def setup_aws_client(config):
...
+                        config=timeout_config   # pass the config to add the request_timeout
+                        )
 
 def get_stream_client(config):
+    # if request_timeout is other than 0,"0" or "" then use request_timeout
+    request_timeout = config.get('request_timeout')
+    if request_timeout and float(request_timeout):
+        request_timeout = float(request_timeout)
+    else: # If value is 0,"0" or "" then set default to 300 seconds.
+        request_timeout = REQUEST_TIMEOUT
```
Comment (savan-chovatiya) [inline tap_dynamodb/dynamodb.py:95]: @namrata270998, Create a separate function to calculate request timeout based on provided config as we are calculating the same thing in multiple functions.

File: tap_dynamodb/sync_strategies/log_based.py [hunk 2]
```diff
@@ -212,7 +216,10 @@ def has_stream_aged_out(state, table_name):
     # stream then we consider the stream to be aged out
     return time_span > datetime.timedelta(hours=19, minutes=30)
 
-
+@backoff.on_exception(backoff.expo,
+                          (ReadTimeoutError, ConnectTimeoutError),
+                          max_tries=5,
+                          factor=2)
```
Comment (savan-chovatiya) [inline tap_dynamodb/sync_strategies/log_based.py:222]:
```suggestion
@backoff.on_exception(backoff.expo,
                      (ReadTimeoutError, ConnectTimeoutError),
                      max_tries=5,
                      factor=2)
```

File: tap_dynamodb/sync_strategies/log_based.py [hunk 3]
```diff
@@ -115,7 +115,11 @@ def sync_shard(shard, seq_number_bookmarks, streams_client, stream_arn, projecti
     singer.write_state(state)
     return rows_synced
 
-
+# Backoff for both ReadTimeout and ConnectTimeout error for 5 times
+@backoff.on_exception(backoff.expo,
+                          (ReadTimeoutError, ConnectTimeoutError),
+                          max_tries=5,
+                          factor=2)
```
Comment (savan-chovatiya) [inline tap_dynamodb/sync_strategies/log_based.py:122]:
```suggestion
@backoff.on_exception(backoff.expo,
                      (ReadTimeoutError, ConnectTimeoutError),
                      max_tries=5,
                      factor=2)
```

File: tap_dynamodb/sync_strategies/full_table.py [hunk 4]
```diff
@@ -35,7 +37,11 @@ def scan_table(table_name, projection, last_evaluated_key, config):
 
         has_more = result.get('LastEvaluatedKey', False)
 
-
+# Backoff for both ReadTimeout and ConnectTimeout error for 5 times
+@backoff.on_exception(backoff.expo,
+                          (ReadTimeoutError, ConnectTimeoutError),
+                          max_tries=5,
+                          factor=2)
```
Comment (savan-chovatiya) [inline tap_dynamodb/sync_strategies/full_table.py:44]:
```suggestion
@backoff.on_exception(backoff.expo,
                      (ReadTimeoutError, ConnectTimeoutError),
                      max_tries=5,
                      factor=2)
```

File: .circleci/config.yml [hunk 5]
```diff
@@ -18,7 +18,18 @@ jobs:
           name: 'pylint'
           command: |
             source /usr/local/share/virtualenvs/tap-dynamodb/bin/activate
-            make lint
+            pylint tap_dynamodb --disable 'duplicate-code,missing-module-docstring,missing-class-docstring,too-few-public-methods,line-too-long,fixme,missing-function-docstring,raise-missing-from,consider-using-f-string,too-many-locals,invalid-name,too-many-arguments'
```
Comment (savan-chovatiya) [inline .circleci/config.yml:21]: @namrata270998 Do we need this change for any specific reason?.

File: tap_dynamodb/sync_strategies/log_based.py (hunk 3 above)
Comment (namrata270998) [reply tap_dynamodb/sync_strategies/log_based.py]: done

File: tap_dynamodb/dynamodb.py (hunk 1 above)
Comment (namrata270998) [reply tap_dynamodb/dynamodb.py]: done

File: tap_dynamodb/sync_strategies/log_based.py (hunk 2 above)
Comment (namrata270998) [reply tap_dynamodb/sync_strategies/log_based.py]: done

File: .circleci/config.yml (hunk 5 above)
Comment (namrata270998) [reply .circleci/config.yml]: reverted the change

**Review by dbshah1212 — APPROVED** [review]

**Review by karanpanchal-crest — CHANGES_REQUESTED** [review]

File: tests/unittests/test_request_timeout.py [hunk 6]
```diff
@@ -0,0 +1,217 @@
+from tap_dynamodb.sync_strategies import full_table, log_based
+from tap_dynamodb import LOGGER, discover, dynamodb
+import tap_dynamodb
+import unittest
+from unittest import mock
+from unittest.mock import Mock
+from unittest.case import TestCase
+from botocore.exceptions import ClientError, ConnectTimeoutError, ReadTimeoutError
```
Comment (karanpanchal-crest) [inline tests/unittests/test_request_timeout.py:8]: Remove `ClientError` if not used

File: Makefile [hunk 7]
```diff
@@ -4,6 +4,6 @@ lint-tests:
...
 
 lint-code:
-	pylint tap_dynamodb -d broad-except,chained-comparison,empty-docstring,fixme,invalid-name,line-too-long,missing-class-docstring,missing-function-docstring,missing-module-docstring,no-else-raise,no-else-return,too-few-public-methods,too-many-arguments,too-many-branches,too-many-lines,too-many-locals,ungrouped-imports,wrong-spelling-in-comment,wrong-spelling-in-docstring,raise-missing-from,consider-using-f-string
+	pylint tap_dynamodb -d broad-except,chained-comparison,empty-docstring,fixme,invalid-name,line-too-long,missing-class-docstring,missing-function-docstring,missing-module-docstring,no-else-raise,no-else-return,too-few-public-methods,too-many-arguments,too-many-branches,too-many-lines,too-many-locals,ungrouped-imports,wrong-spelling-in-comment,wrong-spelling-in-docstring,raise-missing-from,consider-using-f-string,duplicate-code
```
Comment (karanpanchal-crest) [inline Makefile:7]: Remove the `duplicate-code` from the global level. Instead handle it where needed

**Review by karanpanchal-crest — CHANGES_REQUESTED** [review]

File: tap_dynamodb/sync_strategies/full_table.py [hunk 8]
```diff
@@ -35,7 +37,11 @@ def scan_table(table_name, projection, last_evaluated_key, config):
 
         has_more = result.get('LastEvaluatedKey', False)
 
-
+# Backoff for both ReadTimeout and ConnectTimeout error for 5 times
```
[the remark this answers is not in the document]
Comment (namrata270998) [reply tap_dynamodb/sync_strategies/full_table.py:40]: @karanpanchal-crest The function `scan_table` contains a `yield`. Hence the backoff won't work on that function

File: Makefile (hunk 7 above)
Comment (namrata270998) [reply Makefile]: done

File: tests/unittests/test_request_timeout.py (hunk 6 above)
Comment (namrata270998) [reply tests/unittests/test_request_timeout.py]: removed unused imports

**Review by karanpanchal-crest — APPROVED** [review]

File: tests/unittests/test_request_timeout.py (hunk 6 above)
Comment (namrata270998) [reply tests/unittests/test_request_timeout.py]: done

File: tap_dynamodb/dynamodb.py [hunk 9]
```diff
@@ -65,17 +67,42 @@ def setup_aws_client(config):
...
-    return boto3.client('dynamodb', config['region_name'])
+                            region_name=config['region_name'],
+                            config=timeout_config   # pass the config to add the request_timeout
+                            )
+    return boto3.client('dynamodb', config['region_name'],
+                        config=timeout_config   # pass the config to add the request_timeout
+                        )
 
 def get_stream_client(config):
+    # get the request_timeout
```
Comment (KrisPersonal) [inline tap_dynamodb/dynamodb.py:95]: What is the difference between get_stream_client and get_client function?

File: tests/unittests/test_request_timeout.py [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+
+    @mock.patch('boto3.client')
+    @mock.patch("tap_dynamodb.dynamodb.Config")
+    def test_config_provided_string_request_timeout(self, mock_config, mock_client):
+        """ 
+            Unit tests to ensure that request timeout is set based on config string value
+        """
+        config = {"region_name": "dummy_region", "request_timeout": "100"}
+        dynamodb.get_client(config)
+        mock_config.assert_called_with(connect_timeout=100, read_timeout=100)
```
[the remark this answers is not in the document]
Comment (namrata270998) [reply tests/unittests/test_request_timeout.py:157]: @kspeer825 updated the code and replaced values with reusable variables.

File: tap_dynamodb/dynamodb.py (hunk 9 above)
Comment (namrata270998) [reply tap_dynamodb/dynamodb.py:95]: @KrisPersonal Amazon DynamoDB Streams provides API actions for accessing streams and processing stream records, which can be done through `get_stream_client`. Whereas, the `get_client` returns a low-level client representing Amazon DynamoDB

File: .circleci/config.yml [hunk 11]
```diff
@@ -19,6 +19,17 @@ jobs:
           command: |
             source /usr/local/share/virtualenvs/tap-dynamodb/bin/activate
             make lint
+      - run:
```
[the remark this answers is not in the document]
Comment (namrata270998) [reply .circleci/config.yml:25]: @kspeer825 Updated this file as requested

File: .circleci/config.yml [hunk 12]
```diff
@@ -19,19 +22,26 @@ jobs:
...
+            nosetests --with-coverage --cover-erase --cover-package=tap_dynamodb --cover-html-dir=htmlcov tests/unittests
+            coverage html
+      - store_test_results:
+          path: test_output/report.xml
+      - store_artifacts:
+          path: htmlcov
       - run:
           name: 'Tap Tester'
           command: |
             aws s3 cp s3://com-stitchdata-dev-deployment-assets/environments/tap-tester/sandbox tap-tester.env
```
Comment (kspeer825) [inline .circleci/config.yml:39]: These environment variables are not the up-to-date set. Please use `tap-tester-sandbox` over `sandbox`.

File: .circleci/config.yml [hunk 13]
```diff
@@ -19,19 +22,26 @@ jobs:
...
-                     --email=harrison+sandboxtest@stitchdata.com \
-                     --password=$SANDBOX_PASSWORD \
-                     --client-id=50 \
-                     tests
+            run-test --tap=tap-dynamodb tests
+      - slack/notify-on-failure:
+          only_for_branches: master
 workflows:
   version: 2
   commit:
```
Comment (kspeer825) [inline .circleci/config.yml:47]: Can you add the `tap-tester-user` context as well.

**Review by kspeer825 — CHANGES_REQUESTED** [review]
Just a couple config changes. Tests look good.

File: .circleci/config.yml (hunk 13 above)
Comment (namrata270998) [reply .circleci/config.yml:47]: Added the tap-tester-user context

File: .circleci/config.yml (hunk 12 above)
Comment (namrata270998) [reply .circleci/config.yml]: Update to tap-tester-sandbox env variables set

**Review by kspeer825 — APPROVED** [review]

**Pull request merged**